Mathlib/CategoryTheory/LocallyCartesianClosed/ChosenPullback.lean

```diff
@@ -0,0 +1,308 @@
+  let φ₂' : U ⟶ (pullback g).obj (Over.mk f) := Over.homMk φ₂ (by simpa using h₂.symm)
+  have : φ₁' = φ₂' := by
+    apply (adj.homEquiv U _).symm.injective
+    apply (forget X).map_injective
+    simpa using h₁
+  exact congr_arg CommaMorphism.left this
+
+section Lift
+
+variable {W : C} (a : W ⟶ Y) (b : W ⟶ Z) (h : a ≫ f = b ≫ g := by cat_disch)
+
+/-- Given morphisms `a : W ⟶ Y` and `b : W ⟶ Z` satisfying `a ≫ f = b ≫ g`,
+constructs the unique morphism `W ⟶ pullbackObj f g` which lifts `a` and `b`. -/
+def lift : W ⟶ pullbackObj f g :=
+  (((mapPullbackAdj g).homEquiv (Over.mk b) (Over.mk f)) (Over.homMk a)).left
+
+@[reassoc (attr := simp)]
+theorem lift_fst : lift a b h ≫ fst f g = a := by
+  let adj := mapPullbackAdj g
+  let a' : (Over.map g).obj (Over.mk b) ⟶ Over.mk f := Over.homMk a h
+  have : (Over.map g).map (adj.homEquiv (.mk b) (.mk f) (Over.homMk a)) ≫ fst' f g = a' := by
+    simp only [Functor.id_obj, ← Adjunction.homEquiv_counit, Equiv.symm_apply_apply, adj, a']
+  exact congr_arg CommaMorphism.left this
+
+@[reassoc (attr := simp)]
+theorem lift_snd : lift a b h ≫ snd f g = b := by
+  simp [lift]
+
+end Lift
+
+section PullbackMap
+
+variable (f g)
+
+/-- The functoriality of `pullbackObj f g` in both arguments: Given a map from the pullback cospans
+of `f' : Y' ⟶ X'` and `g' : Z' ⟶ X'` to the pullback cospan of `f : Y ⟶ X` and `g : Z ⟶ X`
+as in the diagram below
+```
+Y' ⟶ Y
+  ↘   ↘
+  X' ⟶ X
+  ↗   ↗
+Z' ⟶ Z
+```
+if the morphisms `g'` and `g` both have chosen pullbacks, then we get an induced morphism
+`pullbackMap f g f' g' comm₁ comm₂` from the chosen pullback of
+`f' : Y' ⟶ X'` along `g'` to the chosen pullback of `f : Y ⟶ X` along `g`.
+Here `comm₁` and `comm₂` are the commutativity conditions of the squares in the diagram above.
+-/
+def pullbackMap {Y' Z' X' : C} (f' : Y' ⟶ X') (g' : Z' ⟶ X') [ChosenPullback g']
+    (γ₁ : Y' ⟶ Y) (γ₂ : Z' ⟶ Z) (γ₃ : X' ⟶ X)
+    (comm₁ : f' ≫ γ₃ = γ₁ ≫ f := by cat_disch) (comm₂ : g' ≫ γ₃ = γ₂ ≫ g := by cat_disch) :
+    pullbackObj f' g' ⟶ pullbackObj f g :=
+  lift (fst f' g' ≫ γ₁) (snd f' g' ≫ γ₂)
+    (by rw [assoc, ← comm₁, ← assoc, condition, assoc, comm₂, assoc])
+
+variable {f g}
+
+@[reassoc (attr := simp)]
+theorem pullbackMap_fst {Y' Z' X' : C} {f' : Y' ⟶ X'} {g' : Z' ⟶ X'} [ChosenPullback g']
+    {γ₁ : Y' ⟶ Y} {γ₂ : Z' ⟶ Z} {γ₃ : X' ⟶ X} (comm₁ comm₂ := by cat_disch) :
+    pullbackMap f g f' g' γ₁ γ₂ γ₃ comm₁ comm₂ ≫ fst f g = fst f' g' ≫ γ₁ := by
+  simp only [pullbackMap, lift_fst]
+
+@[reassoc (attr := simp)]
+theorem pullbackMap_snd {Y' Z' X' : C} {f' : Y' ⟶ X'} {g' : Z' ⟶ X'} [ChosenPullback g']
+    {γ₁ : Y' ⟶ Y} {γ₂ : Z' ⟶ Z} {γ₃ : X' ⟶ X} (comm₁ comm₂ := by cat_disch) :
+    pullbackMap f g f' g' γ₁ γ₂ γ₃ comm₁ comm₂ ≫ snd f g = snd f' g' ≫ γ₂ := by
+  simp only [pullbackMap, lift_snd]
+
+@[simp]
+theorem pullbackMap_id : pullbackMap f g f g (𝟙 Y) (𝟙 Z) (𝟙 X) = 𝟙 _ := by
+  apply hom_ext <;> simp
```

Comment:
Untested, but as `hom_ext` is an `ext` lemma, it should reasonably work (and we should investigate how to make this work otherwise!).
```suggestion
  cat_disch
```